WIP: 8239589: JavaFX UI will not repaint after reconnecting via Remote Desktop

kevinrushforth opened the pull request:
This is a WIP fix for:

[JDK-8239589](https://bugs.openjdk.java.net/browse/JDK-8239589): JavaFX UI will not repaint after reconnecting via Remote Desktop

An existing WIP PR #315 by @Schmidor was a good first step in solving this. While reviewing and testing that proposed fix, I noticed two main problems when reconnecting to an RDP connection, as described in [this comment of PR #315](https://github.com/openjdk/jfx/pull/315#issuecomment-777127835):

1. The JavaFX application thread gets an NPE while trying to access the `GraphicsPipeline` (which is intended to be a singleton)
2. Buttons and other UI controls were repainted incorrectly

In order to more easily debug this, I added code to simulate the device removed case that happens when reconnecting to a system over RDP, by reporting `D3DERR_DEVICEREMOVED` every 10 frames.

These are the changes I have made (so far) relative to PR #315 :

1. Replaced the logic that disposed and recreated the `GraphicsPipeline` singleton with logic to reset (and reuse) the existing Java object. The native d3d device object is destroyed and recreated, and the `D3DResourceFactory` and `D3DContext` objects are disposed and recreated (along with their native objects), but the `D3DPipeline` object itself (the concrete subclass of `GraphicsPipeline`) is not.
2. Fixed the gradient cache in `PaintHelper` so it will recreate the gradient textures and offsets when the surface is lost; this was the primary cause of the rendering artifacts. I also found and fixed a couple other places where "permanent" textures were not being checked for a lost surface. There might be more such instances, so I need to do a complete search for them.
3. Implemented the `dispose` method in `D3DResourceFactory` and moved the call to `notifyReleased` from `D3DPipeline` dispose to there.
4. Added a `dispose` method to `D3DContext`.

With the above changes, most things are now working with some issues found during testing:

* When rendering a `WebView`, there is a brief, but noticeable, flash to white when the device is destroyed and recreated
* If media is playing when you disconnect from RDP, it won't repaint properly when reconnecting (this one doesn't fail when simulating it)
* When the device is destroyed and recreated, the `resourceFreed` method occasionally finds an attempt to free an unmanaged resource. This is currently an error, but was changed to a warning by PR #315 -- I had to leave it as a warning, but added back in the exception so we could catch it and print the stack trace.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [ ] Commit message must refer to an issue
- [ ] Change must be properly reviewed

### Download
`$ git fetch https://git.openjdk.java.net/jfx pull/403/head:pull/403`
`$ git checkout pull/403`

kevinrushforth commented:
After I am done debugging it, I intend to close this PR and create a new one with the proposed fix.

kevinrushforth commented:
@Schmidor @arapte This is not ready for formal review, but you can take a look if you want.

kevinrushforth commented:
This PR contains the original commits from PR #315 so that anyone who wants to see the diffs between that PR and this one, can easily do so by just looking at the diffs for [commit d0ba1d4](https://github.com/openjdk/jfx/pull/403/commits/d0ba1d4f319983bc1f2018e601045c32016191dc).

kevinrushforth commented on modules/javafx.graphics/src/main/java/com/sun/prism/d3d/D3DContext.java:
```diff
@@ -141,13 +145,59 @@ private void setLost() {
 
+        // KCR: debug
+        // KCR: Periodically simulate device removed state
+//        hr = simulateDeviceRemoved(hr);
```
If you want to test this on a Windows 10 system without using RDP, you can uncomment this line, which will cause a `D3DERR_DEVICEREMOVED` to be sent every 10 frames.

Schmidor commented on modules/javafx.graphics/src/main/java/com/sun/prism/d3d/D3DPipeline.java:
```diff
@@ -207,7 +237,15 @@ public int getAdapterOrdinal(Screen screen) {
+            // adapters lost, recreate
+            System.err.println("KCR: adapters lost, recreate -- will this ever happen??");
+            D3DPipeline.getInstance().reinitialize();
+            return null;
```
I have those, probably at the beginning of the reconnect to remote desktop:

> D3DContext::dispose
> KCR: reinitialize D3DPipeline
>     reset D3DPipeline
> KCR: adapters lost, recreate -- will this ever happen??
> KCR: reinitialize D3DPipeline
>     reset D3DPipeline
> KCR: adapters lost, recreate -- will this ever happen??
> KCR: reinitialize D3DPipeline
>     reset D3DPipeline
> KCR: adapters lost, recreate -- will this ever happen??
> KCR: reinitialize D3DPipeline
>     reset D3DPipeline
> constructor: com.sun.prism.d3d.D3DContext@29e6e497

kevinrushforth replied:
I see. Thanks for confirming. I wonder whether it will reinitialize correctly without this? Ideally we wouldn't reinitialize (with the associated destruction and recreation of the native d3d device) multiple times like this.

Schmidor replied:
Ok, I tried commenting the reinitialize here out.
That ended in "KCR: adapters lost, recreate -- will this ever happen??" on every event and no recover.

kevinrushforth replied:
OK, looks like it is needed then. Thanks.

kevinrushforth replied:
I would still like to understand this: @Schmidor does this happen every time for you? Is there something you are doing in addition to disconnecting from RDP and reconnecting (e.g., putting the remote device to sleep, unplugging an external monitor, etc)?

Schmidor replied:
Just closing the RDP connection. The errors occur on the reconnect.
The application is a WebView embedded in a JFXPanel in Swing.
Running on HighDPI / 200% on a Windows 10. So there is JFX/Swing interop and scaling involved, if that might influence that issue.

kevinrushforth replied:
I tried that, but still didn't reproduce it. I pushed a couple additional changes, plus some debugging code (in particular, it will give a stack trace of all calls to reinitialize).

If you get a spare minute, can you try it again and attach the log file (or send it to me offline)?

In the mean time, I'll look into the more serious issue, which is the problem with media failing to reconnect.

Schmidor replied:
[rdp-reconnect.txt](https://github.com/openjdk/jfx/files/6039065/rdp-reconnect.txt)
The reset() fails after the first try because factories is null then. I'm adding a check in the loop there and try again.

Schmidor replied:
After eliminating the NPE
[rdp-reconnect2.txt](https://github.com/openjdk/jfx/files/6039263/rdp-reconnect2.txt)

kevinrushforth replied:
OK. The interesting thing is:

> Error: unable to reinitialize D3D graphics device

This means that the native call to `nInit` failed for some reason. I'm interested in knowing why `nInit` fails. Can you run it with the following env variable and system property set?

```
export NWT_TRACE_LEVEL=3
java -Dprism.verbose=true ...
```

It will be pretty noisy.

Schmidor replied:
[rdp-reconnect3.txt](https://github.com/openjdk/jfx/files/6039443/rdp-reconnect3.txt)

kevinrushforth replied:
Thanks for doing the test run and providing the log. The following shows the error:

```
        (I) D3DPipeline_nInit
        (I) D3DPPLM::InitAdapters()
        (I) CheckAdaptersInfo
        (I) ------------------
----->  (E) D3DPPLM::CheckAdaptersInfo: no suitable adapters found
----->  (E) InitD3D: failed to init adapters
        (I) D3DPPLM::ReleaseD3D()
        (I) D3DPPLM::ReleaseAdapters()
        (W) Release: resource pd3d9=0x2d38fba0 not released: refs=1
        (I) D3DPipeline_nDispose
        ...
        D3DPipelineManager: Created D3D9Ex device
        Error: unable to reinitialize D3D graphics device
```

This is the first call to `nInit` after we get `D3DERR_DEVICEREMOVED`.

I'll dig further.

kevinrushforth commented:
I did additional debugging and determined that the `WebView` repaint issue I mentioned in the description is not a problem. It is working as designed; the flashing only happens when I simulate the device lost state, and is not seen in actual use.

I also tested a few other cases during RDP reconnect, including animation, Swing interop (both `JFXPanel` and `SwingNode`), and SWT interop (`FXCanvas`), all of which work. Based on my testing so far, the following issues remain:

* If media is playing when you disconnect from RDP, it won't continue playing you reconnect (this one doesn't fail when simulating it, so might be something in `DirectShow`, which is used by `jfxmedia` on Windows).
* When the device is destroyed and recreated, the `resourceFreed` method occasionally finds an attempt to free an unmanaged resource. As noted in the description, this is was changed from an error to a warning. I see it occasionally after reinitializing the device, both on the real system and when simulating device removal.

Additionally, I would like to know why the check for a device with 0 adapters is needed in some situations, as reported above (I haven't ever seen this in my testing, so I can't say what is causing it). The check for that case seems safe enough, so even without knowing how it can happen, it isn't a blocker.

kevinrushforth commented:
While doing some stress testing, using the simulated device removal logic set to reinitialize every 10 frames, I ran into what looks like a memory leak. I left an animating app running for quite a while, which is enough to go through thousands of calls to dispose / initialization steps, and after a while it got an OOM error. We are also likely leaking native resources.

I should note that the number of reinitialization calls is more than would be expected in several months of a continuously-running app on a real system, but I still want to run the problem to ground.

kevinrushforth commented:
Running with `-verbose:gc` shows just how big the leak is. I will definitely need to track this down before sending the real PR. I'll likely look at it in conjunction with the `resourceFreed` assertion failure, since both point to a problem with resource management in the face of device reinitialization.

kevinrushforth commented:
I just pushed three new commits (in addition to merging the latest changes from master) with the following changes:

* Decora PPSRenderer: defer checking the resource factory until we are rendering (it was accessing it on the wrong thread), and handle the state where it is not yet ready
* Added a couple more disposal checks in WebView rendering code to avoid exceptions while decoding; handle exception in PrismInvoker to avoid WebView crash if RenderJob throws exception
* A few more miscellaneous disposal fixes (all should be done now)
* Debugging cleanup: Removed, qualified with verbose, or turned into log messages


During the course of my testing I discovered a couple more issues, in addition to the ones already known. Here is a complete list of outstanding work:

1. 3D shapes, specifically: D3DMesh, D3DMeshView, and D3DPhongMaterial do not handle device removal / disposal at all. They will fail upon device reconnect, either by not rendering anything or by crashing with a null pointer access in native code. This must be fixed before the actual review goes out.
2. WebView: there is a possible memory leak when device isn't ready after first reset, WCRenderQueueImpl::gc held in JNIGLobal (possibly for follow-up issue). This doesn't happen often, so we might be able to defer this to a follow-on issue.
3. Canvas doesn't preserve the contents after a device reconnect (noticed while running Zoomy, where the bg color is wrong after device reinitialization). This might point to a need to let the app know they have to repaint, since there is no possible way to preserve the contents of the texture when the device is lost. This will likely need to be a follow-up issue
4. There are still a couple remaining FIXME comments that need to be evaluated / addressed
5. I still need to resolve the problem of returning resources to texture pool, although my initial evaluation is that we can defer this to a follow-up issue
6. Media: all looks good, except for the problem mentioned earlier where a media stream playing at the time of a reconnect doesn't continue playing. Reloading the media works fine. This is not directly related to this bug, so will be handled as a follow-up issue.

kevinrushforth commented:
I just created the final PR #430 to fix this bug. As a result, I'm closing this PR
